### PR TITLE
Auto highlight selected

### PR DESCRIPTION
**Pull request by tris203**
#396 but to the right branch
Adds request in #384

### CODE REVIEW

**Comment by tris203** [comment]
@willothy thoughts?

File: lua/harpoon/ui.lua [hunk 1]
```diff
@@ -126,6 +135,13 @@ function HarpoonUI:toggle_quick_menu(list)
 
     local contents = self.active_list:display()
     vim.api.nvim_buf_set_lines(self.bufnr, 0, -1, false, contents)
+    if self.settings.select_current then
+        for idx, line in ipairs(contents) do
+            if line == currentFileName then
+                vim.api.nvim_win_set_cursor(self.win_id, { idx, 0 })
+            end
```
Comment (willothy) [inline lua/harpoon/ui.lua:142]: You are setting the cursor, but not applying highlights. I would suggest to move the if statement inside the loop:

```lua
for idx, line in ipairs(contents) do
  if line == currentFileName  then
    vim.api.nvim_buf_add_highlight(...)
    if self.settings.select_current then
      vim.api.nvim_win_set_cursor(...)
    end
    break -- also make sure to break after you find the line
  end
end
```

File: lua/harpoon/utils.lua [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
 
+function M.normalize_path(buf_name, root)
+    return Path:new(buf_name):make_relative(root)
+end
+
 function M.trim(str)
-   return str:gsub("^%s+", ""):gsub("%s+$", "")
+    return str:gsub("^%s+", ""):gsub("%s+$", "")
 end
+
```
Comment (willothy) [inline lua/harpoon/utils.lua:11]: You should run this thorough `stylua`

**Review by willothy — COMMENTED** [review]
Looks good to me so far, but you're not setting the highlight, just the cursor :)

File: lua/harpoon/utils.lua (hunk 2 above)
Comment (tris203) [reply lua/harpoon/utils.lua:11]: I have

File: lua/harpoon/ui.lua (hunk 1 above)
Comment (tris203) [reply lua/harpoon/ui.lua:145]: Done

**Review by tris203 — COMMENTED** [review]
Ready for review

File: lua/harpoon/utils.lua [hunk 3]
```diff
@@ -1,8 +1,14 @@
+local Path = require("plenary.path")
 local M = {}
 
+function M.normalize_path(buf_name, root)
+    return Path:new(buf_name):make_relative(root)
+end
+
 function M.trim(str)
-   return str:gsub("^%s+", ""):gsub("%s+$", "")
+    return str:gsub("^%s+", ""):gsub("%s+$", "")
```
Comment (willothy) [inline lua/harpoon/utils.lua:9]: lgtm! Except for this singular extra space here lol

Comment (tris203) [reply lua/harpoon/utils.lua:9]: 4 space indenting, thats where its meant to be

Comment (willothy) [reply lua/harpoon/utils.lua:9]: LOL idk how I didn't see that, my bad.

**Comment by tris203** [comment]
I assume, given the direction that this will need adjusting to pull the logic out of UI

I would guess the best way is to make toggle_quick_menu() take a second optional param of CurrentItem which is used for the highlighting/focus logic

Then a function passed back from the API get_current() which contains the logic for determining the name of the current item

**Comment by ThePrimeagen** [comment]
so i am going to be removing all highlighting from Harpoon and i am going to start emitting out events with win/buf id, and perhaps some extra state such that you can add whatever highlights you wish

i don't want harpoon having too much logic.  So i am "likely" not going to accept any PRs that add logic but instead adding hooks

**Comment by willothy** [comment]
Actually this does raise an important point (even if the highlighting is removed from this PR). Not sure how I didn't think of this before, but what if the list items aren't paths? I think that if there is going to be a way to determine the current item, it needs to be a method on the List Config so that different lists can modify the default behavior. It also needs to be optional because for some lists (commands, for example) there is no "current" item, so it would not make sense to show one.

**Comment by ThePrimeagen** [comment]
@willothy there is equals and display

From the UI we call display, else we call equals

**Comment by willothy** [comment]
Yes, sorry I should've been more clear. I mean that we're always checking `nvim_get_current_buf` at the moment so even if we do use `equals` and `display`, we're still only able to check against lists of buffers because the current thing in `equals(list_item, current)` will always be the name of the current buffer.

**Comment by ThePrimeagen** [comment]
yes, these things can be tricky

to me its an information problem.  perhaps we could provide "from_buffer" or some option to let you know whence you came

**Comment by willothy** [comment]
> to me its an information problem. perhaps we could provide "from_buffer" or some option to let you know whence you came

Yes, that's perfect. A more generic way to check "is this buffer related to this item" than checking the name.

**Comment by pockata** [comment]
I was just about to rewrite in Lua that SOB that I PR'd in #199 to drop the `vim.cmd` and `string.format` calls, but I noticed this PR :) 

I just wanted to point out that with `vim.api.nvim_buf_add_highlight` when you `dd` the highlighted entry and paste it again it will lose the highlighting. That's why the better approach is to use `vim.fn.matchadd` (perhaps `vim.fn.matchaddpos` too, but I haven't used it) which preserves the highlighting after an edit (technically reapplies it automatically when a string matches the pattern). 

`HarpoonCurrentFile` isn't widely used because you have to dig into the source code to find it. That's bad marketing on my behalf. If it was present in the README, for example, I believe this would have been a fan favorite feature. An incredibly sensible default of `highlight link HarpoonCurrentFile String` is an amazing way to spot the current item in the list `blazingly fast`, especially if you're a bit dyslexic. Having the highlighting in a hook which you grab from a "Recipes" wiki page is also a good way to go, but having it as a built-in default will provide a much better experience for new and experienced users alike.

I suggest you add the highlight command in your config, test drive it for a couple of days and then remove it to see if you miss it :)

**Comment by ThePrimeagen** [comment]
Coming back to this i think i want to go a different direction.

i would like to think about how to make this a extension not something inside of harpoon.  If you have an idea for an event to emit where you could hook in and do this work yourself, please do.  I would have no problems with the harpoon-colors plugin doing something for me.  I just don't want to maintain that :)

**Comment by ThePrimeagen** [comment]
if you look at `harpoon/init.lua` and search for this line `self._extensions:emit(Extensions.event_names.LIST_CREATED, list)` you can do the same thing for hooking into the buf / window created.

**Comment by ThePrimeagen** [comment]
I am planning on doing another harpoon season coming up. 

I just feel pulled in every direction, it's very difficult, but I'm very excited 

I will make sure this gets somewhat resolved

**Comment by tris203** [comment]
@ThePrimeagen I can reopen the PR if you think it is still useful?

It was your other direction comment made me think it was no longer relevant

**Comment by pockata** [comment]
This is already possible with the events system. I've created a [small plugin](https://github.com/pockata/harpoon-highlight-current-file) that adds this functionality. The actual code is so small that you can copy it to your config and save yourself a dependency.

**Pull request closed without merging**